# Assignment 2 question 1
Grayson Hineline

Note for portfolio: This assignment was conducted when I was very new to Python and the Pandas library. Given my current skills I would have conducted the assignment much more efficiently.

## Question 1: Determine which of the provided cities was the rainiest in 2020.

In [1]:
import pandas as pd

First, we have to get all the the data from all of the cities into our notebook. I did `low_memory = False` because I'm not doing this analysis on a toaster and I can easily fit everything.

In [2]:
Phoenix = pd.read_csv("AZ_Phoenix.csv", low_memory = False)
LosAngeles = pd.read_csv("CA_LosAngeles.csv", low_memory = False)
SanFran = pd.read_csv("CA_SanFrancisco.csv", low_memory = False)
Denver = pd.read_csv("CO_Denver.csv", low_memory = False)
Jacksonville = pd.read_csv("FL_Jacksonville.csv", low_memory = False)
Atlanta = pd.read_csv("GA_Atlanta.csv", low_memory = False)
Chicago = pd.read_csv("IL_Chicago.csv", low_memory = False)
Boston = pd.read_csv("MA_Boston.csv", low_memory = False)
Minneapolis = pd.read_csv("MN_MinneapolisStPaul.csv", low_memory = False)
Ithaca = pd.read_csv("NY_Ithaca.csv", low_memory = False)
NewYork = pd.read_csv("NY_NewYorkJFK.csv", low_memory = False)
Portland = pd.read_csv("OR_Portland.csv", low_memory = False)
Philidelphia = pd.read_csv("PA_Philadelphia.csv", low_memory = False)
Dallas = pd.read_csv("TX_DallasFtWorth.csv", low_memory = False)
Saltlake = pd.read_csv("UT_SaltLake.csv", low_memory = False)
WashingtonD = pd.read_csv("VA_WashingtonDulles.csv", low_memory = False)
Bellingham = pd.read_csv("WA_Bellingham.csv", low_memory = False)
Seattle = pd.read_csv("WA_Seattle.csv", low_memory = False)

Now we're going to stor all of the data for each of the cities in a list that will be called upon later on.

In [3]:
cities_list = [Phoenix, LosAngeles, SanFran, Denver, Jacksonville, Atlanta,
              Chicago, Boston, Minneapolis, Ithaca, NewYork, Portland, Philidelphia,
              Dallas, Saltlake, WashingtonD, Bellingham, Seattle]


For reference, here is the index of each city:
* 0: Phoenix, 
* 1: LosAngeles, 
* 2: SanFran, 
* 3: Denver, 
* 4: Jacksonville, 
* 5: Atlanta,
* 6: Chicago, 
* 7: Boston, 
* 8: Minneapolis, 
* 9: Ithaca, 
* 10: NewYork, 
* 11: Portland, 
* 12: Philidelphia,
* 13: Dallas, 
* 14: Saltlake, 
* 15: WashingtonD, 
* 16: Bellingham,
* 17: Seattle

Now that we have all the data, let's focus on answering how "rainy" was one of these cities. After we figure out how to do that, we can apply the same stuff to all of the other cities. I have *more than a feeling* that Boston will be a good first city to do, so I'll start with Boston

In [4]:
Boston.info(verbose = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13742 entries, 0 to 13741
Data columns (total 124 columns):
 #    Column                                        Dtype  
---   ------                                        -----  
 0    STATION                                       int64  
 1    DATE                                          object 
 2    REPORT_TYPE                                   object 
 3    SOURCE                                        int64  
 4    AWND                                          float64
 5    BackupDirection                               object 
 6    BackupDistance                                float64
 7    BackupDistanceUnit                            object 
 8    BackupElements                                object 
 9    BackupElevation                               int64  
 10   BackupElevationUnit                           float64
 11   BackupEquipment                               object 
 12   BackupLatitude                              

There's a few ways we could go about figuring the "raininess" of Boston. We could do total rainfall, average rainfall per amount of time, amount of days with rain, etc. When I think of how rainy a place is, I don't think of how "hard" it usually rains, but rather if it is rainying at all. Whether it is pouring rain or just drizzling, I am going to stay inside regardless. So for this, we're going to figure out how many hours of the year it rained in Boston in 2020. Then we can extrapolate that methodology to all the other cities to see which city had the most hours with rain.

Goal: Get a series that contains a True values for all hours that it rained at all.

First we have to get all of the data for all the hours that it rained in the year. To make sure we're not counting hours twice, I want to make sure that we have one observation per hour.

To do that, I have to convert the date column to datetime format.

In [5]:
Boston["DATE"] = pd.to_datetime(Boston["DATE"])
Boston.sort_values(by = "DATE")

STATION                DATE REPORT_TYPE  SOURCE  AWND  \
0      72509014739 2020-01-01 00:54:00       FM-15       7   NaN   
1      72509014739 2020-01-01 01:00:00       FM-12       4   NaN   
2      72509014739 2020-01-01 01:54:00       FM-15       7   NaN   
3      72509014739 2020-01-01 02:54:00       FM-15       7   NaN   
4      72509014739 2020-01-01 03:54:00       FM-15       7   NaN   
...            ...                 ...         ...     ...   ...   
13737  72509014739 2020-12-31 22:00:00       FM-12       4   NaN   
13738  72509014739 2020-12-31 22:54:00       FM-15       7   NaN   
13739  72509014739 2020-12-31 23:54:00       FM-15       7   NaN   
13740  72509014739 2020-12-31 23:59:00       SOD         6   NaN   
13741  72509014739 2020-12-31 23:59:00       SOM         6  11.9   

      BackupDirection  BackupDistance BackupDistanceUnit BackupElements  \
0                 ESE             2.6                 mi           SNOW   
1                 ESE             2.6                 mi           SNOW   
2                 ESE             2.6                 mi           SNOW   
3                 ESE             2.6                 mi           SNOW   
4                 ESE             2.6                 mi           SNOW   
...               ...             ...                ...            ...   
13737             ESE             2.6                 mi           SNOW   
13738             ESE             2.6                 mi           SNOW   
13739             ESE             2.6                 mi           SNOW   
13740             ESE             2.6                 mi           SNOW   
13741             ESE             2.6                 mi           SNOW   

       BackupElevation  ...  ShortDurationPrecipitationValue060  \
0                   20  ...                                 NaN   
1                   20  ...                                 NaN   
2                   20  ...                                 NaN   
3                   20  ...                                 NaN   
4                   20  ...                                 NaN   
...                ...  ...                                 ...   
13737               20  ...                                 NaN   
13738               20  ...                                 NaN   
13739               20  ...                                 NaN   
13740               20  ...                                 NaN   
13741               20  ...                                0.45   

      ShortDurationPrecipitationValue080  ShortDurationPrecipitationValue100  \
0                                    NaN                                 NaN   
1                                    NaN                                 NaN   
2                                    NaN                                 NaN   
3                                    NaN                                 NaN   
4                                    NaN                                 NaN   
...                                  ...                                 ...   
13737                                NaN                                 NaN   
13738                                NaN                                 NaN   
13739                                NaN                                 NaN   
13740                                NaN                                 NaN   
13741                               0.49                                0.56   

       ShortDurationPrecipitationValue120 ShortDurationPrecipitationValue150  \
0                                     NaN                                NaN   
1                                     NaN                                NaN   
2                                     NaN                                NaN   
3                                     NaN                                NaN   
4                                     NaN                                NaN   
...                                   ...      

This is tough because it seems like their are multiple observations per hour. Let's look at just one day then to find a pattern.

In [6]:
mask = (Boston["DATE"] >= "2020-01-05") & (Boston["DATE"] < "2020-01-06")
jan1st = Boston[mask]
jan1st[["DATE", "HourlyPrecipitation", "HourlyPresentWeatherType"]]

DATE HourlyPrecipitation    HourlyPresentWeatherType
141 2020-01-05 00:54:00                   T              -RA:02 |RA |RA
142 2020-01-05 00:57:00                   T              -RA:02 |RA |RA
143 2020-01-05 01:00:00                 NaN                        ||RA
144 2020-01-05 01:54:00                0.04              -RA:02 |RA |RA
145 2020-01-05 02:54:00                0.04  -RA:02 SN:03 |RA RA SN |RA
146 2020-01-05 03:54:00                0.02  -SN:03 RA:02 |RA RA SN |RA
147 2020-01-05 04:00:00                 NaN                         NaN
148 2020-01-05 04:35:00                0.02  -RA:02 SN:03 |RA RA SN |RA
149 2020-01-05 04:54:00                0.01                -SN:03 |SN |
150 2020-01-05 05:34:00                   T                -SN:03 |SN |
151 2020-01-05 05:54:00                   T                -SN:03 |SN |
152 2020-01-05 06:11:00                   T                -SN:03 |SN |
153 2020-01-05 06:54:00                   T                         NaN
154 2020-01-05 07:00:00                 NaN                         NaN
155 2020-01-05 07:54:00                0.00                         NaN
156 2020-01-05 08:54:00                0.00                         NaN
157 2020-01-05 09:54:00                0.00                         NaN
158 2020-01-05 10:00:00                 NaN                         NaN
159 2020-01-05 10:54:00                0.00                         NaN
160 2020-01-05 11:54:00                0.00                         NaN
161 2020-01-05 12:54:00                0.00                         NaN
162 2020-01-05 13:00:00                 NaN                         NaN
163 2020-01-05 13:54:00                0.00                         NaN
164 2020-01-05 14:54:00                0.00                         NaN
165 2020-01-05 15:54:00                0.00                         NaN
166 2020-01-05 16:00:00                 NaN                         NaN
167 2020-01-05 16:54:00                0.00                         NaN
168 2020-01-05 17:54:00                0.00                         NaN
169 2020-01-05 18:54:00                0.00                         NaN
170 2020-01-05 19:00:00                 NaN                         NaN
171 2020-01-05 19:54:00                0.00                         NaN
172 2020-01-05 20:54:00                0.00                         NaN
173 2020-01-05 21:54:00                0.00                         NaN
174 2020-01-05 22:00:00                 NaN                         NaN
175 2020-01-05 22:54:00                0.00                         NaN
176 2020-01-05 23:54:00                0.00                         NaN
177 2020-01-05 23:59:00                 NaN                         NaN

After looking at multiple days, it seems like there isn't really a predictable pattern for when the measurements are taken, and there are often far more than 24 observations per day.

To solve this, I'm going to make a loop that looks through each hour to see if it rained at all. The column I am going to use to do so will be "HourlyPresentWeatherType", because the "HourlyPrecipitation" column doesn't distinguish between types of precipitation (i.e. snow vs drizzle vs rain). We want to just count rain, which is given by the weather type column as "RA".

I want to make sure I can count the amount of hours that it rained first. To do so, I'm going to see if I can count an hour in which rain was measured twice. (I went into the excel sheet to find the first hour where rain was measured twice in one hour.)

In [7]:
#in this hour "RA" was reported twice
mask = (Boston["DATE"] >= "2020-01-04 12:00:00") & (Boston["DATE"] < "2020-01-04 13:00:00")
Bostonhour = Boston[mask]
Bostonhour[["DATE", "HourlyPresentWeatherType"]]

DATE HourlyPresentWeatherType
119 2020-01-04 12:39:00      -RA:02 BR:1 |RA |RA
120 2020-01-04 12:54:00      -RA:02 BR:1 |RA |RA

In [8]:
#in this hour "RA" was reported twice
hour = 23
mask = (Boston["DATE"] >= pd.Timestamp(2020, 1, 4, hour, 0)) & (Boston["DATE"] < pd.Timestamp(2020, 1, 4, hour, 59)) 
Bostonhour = Boston[mask]
Bostonhour[["DATE", "HourlyPresentWeatherType"]]

DATE HourlyPresentWeatherType
138 2020-01-04 23:38:00           -RA:02 |RA |RA
139 2020-01-04 23:54:00           -RA:02 |RA |RA

In [9]:
#making a mask for weather in rained within that hour
hour_mask = Bostonhour["HourlyPresentWeatherType"].str.contains("RA")
hour_mask

138    True
139    True
Name: HourlyPresentWeatherType, dtype: bool

In [10]:
True in hour_mask.values

True

After running all, we'll do a simple counter to count how many times "RA" shows up once in an hour. Here's an example for the 23rd hour of January 4th that we just did:

In [11]:
count = 0
if(True in hour_mask.values):
        count = count + 1
count

1

Now that we figured out how, we're going to iterate through all hours of the year, and count all of the hours that it rained.

This for loop counts the hours in a day that "RA" (rain) was reported for each hour of the day of January 4th:

In [12]:
count = 0
for hour in range(0,24):
    mask = (Boston["DATE"] >= pd.Timestamp(2020, 1, 4, hour, 0)) & (Boston["DATE"] < pd.Timestamp(2020, 1, 4, hour, 59))
    Bostonhour = Boston[mask]
    hour_mask = Bostonhour["HourlyPresentWeatherType"].str.contains("RA")
    if(True in hour_mask.values):
        count = count + 1
    

count

6

*Rain was reported during 6 different hours on Jan 1, 2020*

This for loop builds upon the previous by adding days:

In [13]:
count = 0
for day in range(1, 32):
    for hour in range(0,24):
        mask = (Boston["DATE"] >= pd.Timestamp(2020, 1, day, hour, 0)) & (Boston["DATE"] < pd.Timestamp(2020, 1, day, hour, 59))
        Bostonhour = Boston[mask]
        hour_mask = Bostonhour["HourlyPresentWeatherType"].str.contains("RA")
        if(True in hour_mask.values):
            count = count + 1
    
count

36

*Rain was reported during 36 different hours in Jan 2020*

Doing every month will be a bit more complicated because each month doesn't have the same amount of days. Here, I made a list of all the months of the year such that a given month's index is equal to the number of the month - 1 (for example, the index of September is 9 - 1 = 8), and by accessing any of the months in that list, the number of days in that month can also be accessed (`length_of_month[8]` gives the number of days in the month of September). I'm first going to test that it works for the month of January:

In [14]:
count = 0
month = 1
length_of_month = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
month_index = month - 1
for day in range(1, length_of_month[month_index]+1):
    for hour in range(0,24):
        mask = (Boston["DATE"] >= pd.Timestamp(2020, 1, day, hour, 0)) & (Boston["DATE"] < pd.Timestamp(2020, 1, day, hour, 59))
        Bostonhour = Boston[mask]
        hour_mask = Bostonhour["HourlyPresentWeatherType"].str.contains("RA")
        if(True in hour_mask.values):
            count = count + 1
    
count

36

Seemed like it worked because it yielded the same count as the previous version of the loop (36 both times). So now we'll just add a loop to iterate through all the months:

In [15]:
count = 0
for month in range(1, 13):
    length_of_month = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    month_index = month - 1
    for day in range(1, length_of_month[month_index]+1):
        for hour in range(0,24):
            mask = (Boston["DATE"] >= pd.Timestamp(2020, month, day, hour, 0)) & (Boston["DATE"] < pd.Timestamp(2020, month, day, hour, 59))
            Bostonhour = Boston[mask]
            hour_mask = Bostonhour["HourlyPresentWeatherType"].str.contains("RA")
            if(True in hour_mask.values):
                count = count + 1
    
count

813

There we have it: **Rain was reported during 813 different hours in Boston in 2020**. Now lets just apply it to the other cities.

In [16]:
cities_list = [Phoenix, LosAngeles, SanFran, Denver, Jacksonville, Atlanta,
              Chicago, Boston, Minneapolis, Ithaca, NewYork, Portland, Philidelphia,
              Dallas, Saltlake, WashingtonD, Bellingham, Seattle]

For reference, here is the index of each city:
* 0: Phoenix, 
* 1: LosAngeles, 
* 2: SanFran, 
* 3: Denver, 
* 4: Jacksonville, 
* 5: Atlanta,
* 6: Chicago, 
* 7: Boston, 
* 8: Minneapolis, 
* 9: Ithaca,
* 10: NewYork, 
* 11: Portland, 
* 12: Philidelphia,
* 13: Dallas, 
* 14: Saltlake, 
* 15: WashingtonD, 
* 16: Bellingham,
* 17: Seattle

First I'm going to test that it works on Phoenix:

In [17]:
city = cities_list[0] #Phoenix has an index of 0
count = 0
city["DATE"] = pd.to_datetime(city["DATE"])

for month in range(1, 13):
    length_of_month = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    month_index = month - 1
    for day in range(1, length_of_month[month_index]+1):
        for hour in range(0,24):
            mask = (city["DATE"] >= pd.Timestamp(2020, month, day, hour, 0)) & (city["DATE"] < pd.Timestamp(2020, month, day, hour, 59))
            cityhour = city[mask]
            hour_mask = cityhour["HourlyPresentWeatherType"].str.contains("RA")
            if(True in hour_mask.values):
                count = count + 1
    
count

85

*Rain was reported in just 85 individual hours in Phoenix in 2020.* This makes sense when compared to Boston. More importantly, it works! Let's apply the loop to every city we have data for and add the totals to a list in the order that they are indexed in `cities_list`.

In [18]:
rain_totals = []
for city_index in range(0, len(cities_list)):
    city = cities_list[city_index]
    count = 0
    city["DATE"] = pd.to_datetime(city["DATE"])

    for month in range(1, 13):
        length_of_month = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
        month_index = month - 1
        for day in range(1, length_of_month[month_index]):
            for hour in range(0,24):
                mask = (city["DATE"] >= pd.Timestamp(2020, month, day, hour, 0)) & (city["DATE"] < pd.Timestamp(2020, month, day, hour, 59))
                cityhour = city[mask]
                hour_mask = cityhour["HourlyPresentWeatherType"].str.contains("RA")
                if(True in hour_mask.values):
                    count = count + 1
    
    rain_totals.append(count)

In [19]:
rain_totals

[85,
 167,
 160,
 158,
 538,
 798,
 597,
 776,
 403,
 758,
 825,
 1133,
 809,
 575,
 269,
 750,
 1098,
 1123]

In [20]:
city_list_with_strings = ["Phoenix", "LosAngeles", "SanFran", "Denver", "Jacksonville", "Atlanta", "Chicago", "Boston", "Minneapolis", "Ithaca", "NewYork", "Portland", "Philidelphia", "Dallas", "Saltlake", "WashingtonD", "Bellingham", "Seattle"]
city_list_with_strings

['Phoenix',
 'LosAngeles',
 'SanFran',
 'Denver',
 'Jacksonville',
 'Atlanta',
 'Chicago',
 'Boston',
 'Minneapolis',
 'Ithaca',
 'NewYork',
 'Portland',
 'Philidelphia',
 'Dallas',
 'Saltlake',
 'WashingtonD',
 'Bellingham',
 'Seattle']

In [21]:
rainy_hours = pd.Series(data = rain_totals, index = city_list_with_strings)
rainy_hours

Phoenix           85
LosAngeles       167
SanFran          160
Denver           158
Jacksonville     538
Atlanta          798
Chicago          597
Boston           776
Minneapolis      403
Ithaca           758
NewYork          825
Portland        1133
Philidelphia     809
Dallas           575
Saltlake         269
WashingtonD      750
Bellingham      1098
Seattle         1123
dtype: int64

In [22]:
rainy_hours.sort_values()

Phoenix           85
Denver           158
SanFran          160
LosAngeles       167
Saltlake         269
Minneapolis      403
Jacksonville     538
Dallas           575
Chicago          597
WashingtonD      750
Ithaca           758
Boston           776
Atlanta          798
Philidelphia     809
NewYork          825
Bellingham      1098
Seattle         1123
Portland        1133
dtype: int64

In [23]:
rainy_hours.describe()

count      18.000000
mean      612.333333
std       343.798847
min        85.000000
25%       302.500000
50%       673.500000
75%       806.250000
max      1133.000000
dtype: float64

In [24]:
rainy_hours.median()

673.5

There you have it: **In 2020, Portland had the most hours in which rain was observed at all (1133), and therefore it is the rainiest city in 2020 in this dataset.** Seattle and Bellingham were close behind with both over 1000 hours in which it rained (Seattle only 10 behind at 1123), while Pheonix had by far the lowest amount of hours with rain at 85 hours. Additionally, the average rain hours for all the cities in this data was 612.33 hours.

Here are a few justifications for decisions I made with this analysis:
* Why were hours with drizzle and snow not included in the count? I left these out because the question was about raininess. Even though it is precipitation, snow is not rain, therefore it was left out. Drizzle was left out however, because I felt that drizzle doesn't have the same effect on a person as rain. When it is just "drizzling" outside, you still may go out without a jacket or umbrella. But if it is raining, you notice. You think "oh, its raining!". Therefore, for deciding which city was the 'rainiest', I went with counting just rain hours instead of including other types of weather.
* Why hours of rain instead of days with rain? Because if days of rain were counted, a city where it typically rains for 12 hours a day may be considered less rainy than a city where it rains for just an hour of the day if the latter city has more days where it rained for that single hour. I consider a city where it rains more hours of the year more rainy than a city where it rains more days of the year.